synapse/storage/databases/main/room.py

```diff
@@ -1182,7 +1200,12 @@ async def get_partial_state_rooms_and_servers(
         for row in rows:
             room_id = row["room_id"]
             server_name = row["server_name"]
-            room_servers.setdefault(room_id, []).append(server_name)
+            # There is a foreign key constraint which enforces that every room_id in
+            # partial_state_rooms_servers appears in partial_state_rooms, so this lookup
+            # should always succeed.
+            # TODO(faster joins): To make this robust we should run both SELECTs in the
+            #   same transaction.
```

Comment:
I left this as a TODO for pragmatism's sake (read: I don't want to read database.py right now). I'm assuming it's very unlikely that a new partial join will have started and completed between the two SELECTs.

Reply:
Note that a transaction only helps you at a `REPEATABLE READ` (or higher) isolation level (which, in fairness, we do use by default, but every so often we talk about reducing it).

I'd just do `room_servers.get(room_id)` and add a check for `None`, to save having to worry about it.

Reply:
Thanks---the comment is naive as written.

(The reference here is https://www.postgresql.org/docs/current/transaction-iso.html .)

Reply:
Many thanks. This is 9e9be2b787